### DISCUSSION TITLE
群消息中的message_id和message_seq有什么关系

### DISCUSSION THREAD

**Discussion by lingyunQAQ** [Mrs4s/go-cqhttp#1787] [Q&A]
rt
群消息中的message_id和message_seq有什么关系
seq看起来是有规律的但是id却不是，所以我觉得seq应该和id之间有啥关系，使用/get_msg有一个真实id和seq也一样的
我希望我可以获取seq为118之前的信息比如seq为117or116
看过源码但我还是不太了解该怎么获取这个ID，难道只是int32和int64的区别？
求dalao帮助！

**Comment by xiangxiangxiong9** [comment]
message_id 似乎是 群ID-消息seq (如果是私聊,则为 发送者ID-消息seq )经过crc32处理后得到的[int32且可能为负数]

**Reply by lingyunQAQ** [reply]
有意思，这个负数是规律的还是...算了，早上我试试

**Reply by xiangxiangxiong9** [reply]
> 有意思，这个负数是规律的还是...算了，早上我试试

生成:
https://github.com/Mrs4s/go-cqhttp/blob/8343db5ae3b311aea1f0c26afff203d9edc20284/db/database.go#L102-L104

调用:
[群聊消息] https://github.com/Mrs4s/go-cqhttp/blob/8343db5ae3b311aea1f0c26afff203d9edc20284/coolq/bot.go#L430-L470
[私聊消息] https://github.com/Mrs4s/go-cqhttp/blob/8343db5ae3b311aea1f0c26afff203d9edc20284/coolq/bot.go#L473-L513

**Reply by lingyunQAQ** [reply]
其实俺用的py写的看不懂go呀QAQ

**Reply by xiangxiangxiong9** [reply]
> 其实俺用的py写的看不懂go呀QAQ

[其实我Python用的也挺多的, Go 只能略微推测出一点...]

**Reply by lingyunQAQ** [reply]
中午好，我翻了一下
db.ToGlobalID(m.GroupCode, reply.ReplySeq)

db.ToGlobalID(reply.Sender, reply.ReplySeq),

int32(crc32.ChecksumIEEE([]byte(fmt.Sprintf("%d-%d", code, msgID))))
}
这几个比较重要吧

**Reply by xiangxiangxiong9** [reply]
> 中午好，我翻了一下 db.ToGlobalID(m.GroupCode, reply.ReplySeq)
> 
> db.ToGlobalID(reply.Sender, reply.ReplySeq),
> 
> int32(crc32.ChecksumIEEE([]byte(fmt.Sprintf("%d-%d", code, msgID)))) } 这几个比较重要吧
```Go
func ToGlobalID(code int64, msgID int32) int32 { 
 	return int32(crc32.ChecksumIEEE([]byte(fmt.Sprintf("%d-%d", code, msgID)))) 
 } 
```
是用来生成的
确实...

**Reply by lingyunQAQ** [reply]
啥意思呢.....T_T

**Reply by lingyunQAQ** [reply]
[ChatGPT]这串代码的意思是，使用IEEE CRC32算法计算出一个32位整数，其中fmt . Sprintf("%d-%d", code, msgID)用来表示一个字符串，该字符串的内容是code和msgID的拼接。int32(xxx)表示将得到的CRC32值转换成一个32位整数。

**Reply by xiangxiangxiong9** [reply]
> 啥意思呢.....T_T

这应该能看懂吧？[python3]
```Python3
from binascii import crc32
from ctypes import c_int32


def ToGlobalID(code: int, msgID: int) -> int:
    return c_int32(crc32(f"{code}-{msgID}".encode())).value


if __name__ == '__main__':
    group_id, message_seq = int(input("请输入群号")), int(input("请输入消息seq"))
    print("message_id:", ToGlobalID(group_id, message_seq))
```

**Reply by lingyunQAQ** [reply]
看懂了
